include/ginkgo/core/solver/multigrid.hpp

```diff
@@ -0,0 +1,417 @@
+     */
+    bool apply_uses_initial_guess() const override { return true; }
+
+    /**
+     * Gets the stopping criterion factory of the solver.
+     *
+     * @return the stopping criterion factory
+     */
+    std::shared_ptr<const stop::CriterionFactory> get_stop_criterion_factory()
+        const
+    {
+        return stop_criterion_factory_;
+    }
+
+    /**
+     * Sets the stopping criterion of the solver.
+     *
+     * @param other  the new stopping criterion factory
+     */
+    void set_stop_criterion_factory(
+        std::shared_ptr<const stop::CriterionFactory> other)
+    {
+        stop_criterion_factory_ = std::move(other);
+    }
+
+    /**
+     * Gets the system operator of the linear system.
+     *
+     * @return the system operator
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Gets the list of MultigridLevel operator.
+     *
+     * @return the list of MultigridLevel opertor
+     */
+    const std::vector<std::shared_ptr<const gko::multigrid::MultigridLevel>>
+    get_mg_level_list() const
+    {
+        return mg_level_list_;
+    }
+
+    /**
+     * Gets the list of pre-smoother operator.
+     *
+     * @return the list of pre-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_pre_smoother_list() const
+    {
+        return pre_smoother_list_;
+    }
+
+    /**
+     * Gets the list of mid-smoother operator.
+     *
+     * @return the list of mid-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_mid_smoother_list() const
+    {
+        return mid_smoother_list_;
+    }
+
+    /**
+     * Gets the list of post-smoother operator.
+     *
+     * @return the list of post-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_post_smoother_list() const
+    {
+        return post_smoother_list_;
+    }
+
+    /**
+     * Gets the operator at the coarsest level.
+     *
+     * @return the coarsest operator
+     */
+    std::shared_ptr<const LinOp> get_coarsest_solver() const
+    {
+        return coarsest_solver_;
+    }
+
+    /**
+     * Get the cycle of multigrid
+     *
+     * @return the cycle
+     */
+    multigrid_cycle get_cycle() const { return cycle_; }
+
+    /**
+     * Set the cycle of multigrid
+     *
+     * @param cycle the new cycle
+     */
+    void set_cycle(multigrid_cycle cycle) { cycle_ = cycle; }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Criterion factories.
+         */
+        std::vector<std::shared_ptr<const stop::CriterionFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(criteria, nullptr);
+
+        /**
+         * MultigridLevel Factory list
+         */
+        std::vector<std::shared_ptr<const gko::LinOpFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(mg_level, nullptr);
+
+        /**
+         * Custom selector (level, matrix)
```

Comment:
This is not clear. How about this?
```suggestion
         * Selector function to return the correct element index in the mg_level vector for any given level index.
```